evergreen/eg_app/consumers.py

```diff
@@ -0,0 +1,238 @@
+
+        # Add to group
+        await self.channel_layer.group_add(self.room_group_name, self.channel_name)
+        await self.accept()
+        print("(GOOD) WebSocket connection accepted successfully!")
+
+    # disconnects connection handler
+    async def disconnect(self, close_code):
+
+        await self.channel_layer.group_discard(self.room_group_name, self.channel_name)
+        print(f"(BAD) WebSocket disconnected with code: {close_code}")
+
+    # updated to all clients
+    async def feed_update(self, event):
+
+        # sending a lovey message to websocket
+        await self.send(
+            text_data=json.dumps(
+                {"type": "feed_update", "posts": event["data"]["posts"]}
+            )
+        )
+        print("Feed update sent successfully!")
+
+    # handle receiving like events!
+    async def receive(self, text_data):
+        try:
+            data = json.loads(text_data)
+            if data["type"] == "like":
+                post_id = data["post_id"]
+                user = self.scope["user"]
+
+                # Update like in database
+                success, likes = await self.update_like(post_id, user)
+
+                if success:
+                    post_data = await self.get_post_data(post_id)
+
+                    # Broadcast the updated likes to ALL clients
+                    await self.channel_layer.group_send(
+                        self.room_group_name,
+                        {
+                            "type": "like_update",
+                            "post_id": post_id,
+                            "likes": post_data["likes"],
+                            "likers_display": post_data["likers_display"],
+                            "has_liked": post_data["has_liked"],
+                        },
+                    )
+            elif data["type"] == "upload_post":
+
+                post = await self.handle_post_upload(data)
+
+                if post:
+                    # BROADCAST TO ALL CLIENTS
+                    posts_data = await self.get_all_posts()
+
+                    await self.channel_layer.group_send(
+                        self.room_group_name,
+                        {
+                            "type": "feed_update",
+                            "data": {"posts": posts_data},
+                        },
+                    )
+
+                    # send upload
+                    await self.send(
+                        text_data=json.dumps(
+                            {
+                                "type": "upload_response",
+                                "status": "success",
+                                "post_id": str(post.id),
+                            }
+                        )
+                    )
+        except Exception as e:
+            print(f"Error in receive: {str(e)}")
+
+    @database_sync_to_async
+    def get_all_posts(self):
+
+        Post = apps.get_model("eg_app", "Post")
+
```

Comment:
This returns a generic `Model`, which results in a lot of type errors when you try to use it. Just delete this line (times 3 for each method) and do `from models import Post` at the top

Reply:
https://docs.djangoproject.com/en/5.1/ref/applications/#django.apps.apps.get_model

This is the django recommended way to get a model, and it does not return a "generic" one.

Reply:
It's fine at runtime, but the return type only has the parent class so it doesn't have any details at ~~compile~~ linting time.

When I read into the docs on models when I was working on that earlier, it just had referencing the model by Python importing. This page doesn't use `get_model` at all https://docs.djangoproject.com/en/5.1/topics/db/models/

I assume that one is just if you don't know exactly what type at whatever we call compile time on an interpreted language

Reply:
hmmm, interesting. Since trying `from models import Post` and `from eg_app.models import Post`  fail to work as intended, due to django being angry that apps are not loaded in yet (for some reason), I will put a pause on this, and revert to original. We can circle back to this issue with a bugfix afterwards, but I will look to other stuff for now.

Reply:
Hm, yeah that's weird. I played around a little bit and couldn't find an immediate solution. I wonder if this file is loaded early in the process for some reason? Or maybe it's something to do trying to use it async? Either way, yeah, table it for now I guess. There's more important stuff to worry about

Reply:
Yes I fixed that with the asgi.py import order. I looked up the problem and found a Reddit page that discussed it and pointed to that documentation. You have to initialize the django asgi application before importing any models that belong to that asgi app so that the app's registry can be populated first (my understanding).